test/AutoDiff/SILOptimizer/closure_specialization.sil

```diff
@@ -0,0 +1,116 @@
+// RUN: %target-sil-opt -test-runner %s -o /dev/null 2>&1 | %FileCheck %s
+
+// REQUIRES: swift_in_compiler
+
+sil_stage canonical
+
+import Builtin
+import Swift
+import SwiftShims
+
+import _Differentiation
+
+// ===================== Gathering callsites and corresponding closures ===================== //
+
+//////////////////////////////
+// Single closure call site //
+//////////////////////////////
+sil @$vjpMultiply : $@convention(thin) (Float, Float, Float) -> (Float, Float) 
+
+sil private @$pullback_f : $@convention(thin) (Float, @owned @callee_guaranteed (Float) -> (Float, Float)) -> Float {
+bb0(%0 : $Float, %1 : $@callee_guaranteed (Float) -> (Float, Float)):
+  %2 = apply %1(%0) : $@callee_guaranteed (Float) -> (Float, Float) // users: %5, %4
+  strong_release %1 : $@callee_guaranteed (Float) -> (Float, Float) // id: %3
+  %4 = tuple_extract %2 : $(Float, Float), 0      // user: %7
+  %5 = tuple_extract %2 : $(Float, Float), 1      // user: %6
+  %6 = struct_extract %5 : $Float, #Float._value  // user: %8
+  %7 = struct_extract %4 : $Float, #Float._value  // user: %8
+  %8 = builtin "fadd_FPIEEE32"(%6 : $Builtin.FPIEEE32, %7 : $Builtin.FPIEEE32) : $Builtin.FPIEEE32 // user: %9
+  %9 = struct $Float (%8 : $Builtin.FPIEEE32)     // users: %11, %10
+  debug_value %9 : $Float, let, name "x", argno 1 // id: %10
+  return %9 : $Float                              // id: %11
+}
+
+// reverse-mode derivative of f(_:)
+sil hidden @$s4test1fyS2fFTJrSpSr : $@convention(thin) (Float) -> (Float, @owned @callee_guaranteed (Float) -> Float) {
+bb0(%0 : $Float):
+  specify_test "closure_specialize_gather_call_sites"
+  // CHECK-LABEL: Specializing closures in function: $s4test1fyS2fFTJrSpSr
+  // CHECK: PartialApply call site:   %8 = partial_apply [callee_guaranteed] %7(%6) : $@convention(thin) (Float, @owned @callee_guaranteed (Float) -> (Float, Float)) -> Float // user: %9
+  // CHECK: Passed in closures:
+  // CHECK: 1.   %6 = partial_apply [callee_guaranteed] %5(%0, %0) : $@convention(thin) (Float, Float, Float) -> (Float, Float) // user: %8
+
```

Comment:
Can we check the result of specialization here? And not just debug logs?

Reply:
The specialization does not happen in this PR. It is happening in [this](https://github.com/jkshtj/swift/pull/1) subsequent PR.